pkg/filter/evaluators.go

```diff
@@ -0,0 +1,142 @@
+package filter
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+)
+
+const (
+	// TypeEqual is a filter type that matches exactly the reference value.
+	TypeEqual = "equal"
+
+	// TypeNotEqual is a filter type that matches when the value is different from the reference value (opposite of TypeEqual).
+	TypeNotEqual = "notequal"
+
+	// TypeRegexp is a filter type that matches the value against a reference regular expression.
+	// The reference value must be a regular expression that can compile.
+	// Works only with strings (anything else will evaluate to false).
+	TypeRegexp = "regexp"
+)
+
+// Evaluator is the interface to provide functions for a filter type.
+type Evaluator interface {
+	// Evaluate determines if two given values match.
+	Evaluate(value interface{}) bool
+}
+
+type equal struct {
+	ref interface{}
+}
+
+func newEqual(reference interface{}) Evaluator {
+	return &equal{
+		ref: convertValues(reference),
+	}
+}
+
+// Evaluate returns whether reference and actual are considered equal.
+// It converts numerical values implicitly before comparison.
+func (e *equal) Evaluate(value interface{}) bool {
+	value = convertValues(value)
+
+	if value == e.ref {
+		return true
+	}
+
+	if e.isNil(value) && e.isNil(e.ref) {
+		return true
+	}
+
+	return false
+}
+
+func (e *equal) isNil(v interface{}) bool {
+	if v == nil {
+		return true
+	}
+
+	value := reflect.ValueOf(v)
+	if (value.Kind() == reflect.Interface || value.Kind() == reflect.Ptr) && value.IsNil() {
+		return true
+	}
+
+	return false
+}
+
+// nolint: gocyclo
+func convertValues(value interface{}) interface{} {
+	switch value := value.(type) {
+	case int:
+		return float64(value)
+	case int8:
```

Comment:
you may put it all in a one line like `case int, int8...:` etc

Reply:
Doing that will result in:
"cannot convert value (variable of type interface{}) to float64 (need type assertion) (typecheck)"

Reply:
I have seen this before, let me check what the actual solution was

Reply:
It looks like we have to keep these separate entries for the proper type assertion.